### PR TITLE
Enable add and mul torch ops for int64 tensors converting them to float32

### LINKED ISSUE #1092: Test toch.add and torch.Tensor.add (int64)
### What feature would you like to be added?

Enable/test torch.add for Q1 target models

### Why is this feature needed?

granite-3.3-8b-instruct models (FMS and Huggingface) need torch.add

### Any alternative solution that you may want to share?

_No response_

### Anything else we need to know?

Original issue https://github.com/torch-spyre/torch-spyre/issues/202

### LINKED ISSUE #1396: Enable/test torch.mul for Q2 target models
### What happened?

Enable/test torch.mul for Q2 target models

### What did you expect to happen?

Ministral-3-14B-Instruct-2512, Mistral-Small-3-2-24B-Instruct-2506, gpt-oss-20b, granite-3-3-8b-instruct, granite-3-3-8b-instruct-fms need torch.mul

### How can we reproduce it?

pytest -n 1 --max-worker-restart=99999 -s -rsapd tests/models/test_model_ops.py -k torch_mul

### Any environmental details we need to know?

**torch-spyre commit:** 3d0787d4344edb319e285277c5a28dfd8fd718ea

### Anything else we need to know?

The original issue https://github.com/torch-spyre/torch-spyre/issues/283

### Relevant log output

```Shell

```

### LINKED ISSUE #1547: Enable/mitigate int64 support for Q2 target models
### What feature would you like to be added?

Support torch ops to take 64-bit integer tensors and/or parameters.
- [torch.add and torch.Tensor.add](https://github.com/torch-spyre/torch-spyre/issues/1092)
- [torch.clamp](https://github.com/torch-spyre/torch-spyre/issues/1137)
- [torch.eq](https://github.com/torch-spyre/torch-spyre/issues/1465)
- [torch.index_copy]
- [torch.lt](https://github.com/torch-spyre/torch-spyre/issues/1350)
- [torch.prod](https://github.com/torch-spyre/torch-spyre/issues/1455)
- [torch.cumsum](https://github.com/torch-spyre/torch-spyre/issues/148)

### Why is this feature needed?

Those torch ops are necessary to support the target models.
- https://github.com/torch-spyre/torch-spyre/issues/695
- https://github.com/torch-spyre/torch-spyre/issues/696
- https://github.com/torch-spyre/torch-spyre/issues/698

### Any alternative solution that you may want to share?

CPU off-load

### Anything else we need to know?

Actual use cases for int64 and their expected range:

| Use case                 | Range                                                   |
|--------------------|--------------------------------------|
| Embedding             | ~200K                                                   |
| Rope position         | context window size = 128K or larger |
| KV cache                | Spyre addressable space                    |
| MOE                        | number of experts ~ 16                       |

### LINKED ISSUE #1878: Draft Proposal: Supporting int64 PyTorch Tensors on Spyre via float32 Representation
This is the proposal to solve https://github.com/torch-spyre/torch-spyre/issues/1547
This is proposal for intermediate solution to run 2Q target models.

## 1. Background and Motivation

PyTorch uses `int64` tensors extensively for:

- Indexing and slicing operations
- Integer arithmetic (add, mul, div, mod)
- Shape and size computations
- Cumulative operations (e.g. `cumsum`, `cumprod`)

### Current `torch-spyre` Implementation

- Maps `int64 → int32` on Spyre  
  - See: `types_mapping.h:65–86`
- Falls back to CPU

This behavior introduces unnecessary CPU fallbacks and performance overheads.

### Proposed Approach

- Map `int64 → float32` on Spyre
- Leveraging float32's ability to exactly represent integers up to 2^24 (16,777,216).
- Avoid CPU fallbacks for the majority of transformer-model workloads

**Key insight:**  
Analysis of transformer models (see issue #1692) shows that `int64` tensor values—excluding internal address tensors—fit within the integer range of `float32`.

---

## 2. Current Usage of `int64` Tensors in PyTorch (Consumers)

`int64` tensors are primarily consumed in the following scenarios.

### 2.1 Index Tensors for Indirect Access Operations

- Used as index tensors in:
  - `gather`
  - `scatter`
  - Similar indirect-access operators
- Require conversion from:
  - **Offset tensors (PyTorch)**  
    → **Address tensors (internal to torch-spyre)**

Potential implications:

- Address tensor values may be large
- Special handling may be required
- CPU fallback may be necessary for address generation

---

### 2.2 Slicing Offsets

- Used in slicing expressions (e.g. `x[a:b]`)
- Key observations:
  - Slice offsets are typically **compile-time resolvable**
  - Runtime integer arithmetic on Spyre is **unnecessary**

---

### 2.3 Explicit Conversion to Floating Point

- User explicitly converts integer tensors:
  - e.g. `x.float()`, `add`, `mul`
- Requires:
  - `int → float` conversion

---

## 3. Key Observations (Issue #1692)

- Except for **address tensors**, all observed `int64` tensor values:
  - Fall within the representable range of `float32`
- Address tensors:
  - Do **not** need to be represented as PyTorch tensors
  - Are an internal implementation detail
  - Can be generated or handled independently (e.g. CPU-side)

---

### 4.1 Core Idea

Represent **`int64` PyTorch tensors as `float32` tensors on Spyre**.

- This is a **representation mapping**
- PyTorch-visible semantics remain `int64`
- Physical storage and computation use `float32`

---

### 4.2 Data Format Mapping Change

https://github.com/torch-spyre/torch-spyre/blob/main/torch_spyre/csrc/types_mapping.h#L65-L86

#### Current Mapping

```cpp
{"int64", {DataFormats::IEEE_INT64, DataFormats::IEEE_INT32}},
```
#### Proposed Mapping
```cpp
{"int64", {DataFormats::IEEE_INT64, DataFormats::IEEE_FP32}},
```
Implication:
• All integer arithmetic operations on Spyre are carried out using float32
￼
### 5.1 Transfer from CPU to Spyre
When calling:
x.to("spyre")

if:
• x.dtype == torch.int64
Then:
• Conversion int64 → float32 occurs on CPU
• Converted tensor is transferred to Spyre as float32
￼
### 5.2 Behavior on Spyre
• x.to(dtype=torch.float32)
    • If x.dtype == int64
    • Implemented internally as a clone
• x.to(dtype=torch.int64)
    • If underlying representation is already float32
    • Implemented internally as a clone
Note:
No actual datatype transformation is required on Spyre.
￼
## 6. Benefits
• Integer arithmetic implemented using existing float32 operators
• Avoids unnecessary CPU fallbacks except where unavoidable:
    • e.g. address tensor generation

### PR DESCRIPTION
**Pull request by imaihal**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

This PR implements support for INT64 arithmetic operations (add, mul, sub, minimum, maximum) in the Spyre backend by converting INT64 operations to FP32, performing the computation, and converting back to INT64.

1. Spyre hardware only supports floating-point operations
2. INT64 is stored as INT32 on Spyre device (via `types_mapping.h`)
3. INT32 arithmetic is not supported on Spyre backend
4. FP32 can represent integers in the range [-2^24, 2^24] with exact precision. (From issue #1692, these range is enough in actual models excluding address calculations)

If DDL for type conversion is not available, CPU fallback is used. Currently no DDL for int64 to fp32 conversion.

#### Which issue(s) this PR is related to:

- #1547
- #1692 

#### Special notes for your reviewer:
<details>
<summary>Implementation Details</summary>

**1. Custom CPU Fallback Op (`customops.py`)**
  - Added `spyre::to_dtype_cpu` custom op that:
  - Moves tensor to CPU
  - Performs dtype conversion
  - Moves back to Spyre device

**2. Lowering Logic (`lowering.py`)**
- **New `to_dtype` lowering**: Handles `prims.convert_element_type` with backend capability checking
  - Returns clone if source and destination dtypes match
  - Falls back to CPU conversion if backend doesn't support the conversion
  - Otherwise uses native backend conversion
  
- **New `with_int64_fallback` helper**: Generic wrapper for INT64 operations
  - Converts INT64 inputs to FP32
  - Executes the operation in FP32
  - Converts output back to INT64 (configurable)
  
- **New lowerings for INT64 operations**:
  - `aten.add.Tensor`
  - `aten.mul.Tensor`
  - `aten.sub.Tensor`
  - `aten.minimum.default`
  - `aten.maximum.default`

##### Future Work

Use DDL to run data conversion on Spyre (requires backend support)
</details>

### CODE REVIEW

File: torch_spyre/_inductor/customops.py [hunk 1]
```diff
@@ -499,3 +499,14 @@ def _constant(
...
+
+@torch.library.custom_op("spyre::to_dtype_cpu", mutates_args=(), device_types="spyre")
+def to_dtype_cpu(input: torch.Tensor, dtype: torch.dtype) -> torch.Tensor:
+    warn_fallback(f"conversion from {input.dtype} to {dtype}")
+    return input.cpu().to(dtype=dtype).to(input.device)
+
+
+@to_dtype_cpu.register_fake
+def _(input: torch.Tensor, dtype: torch.dtype) -> torch.Tensor:
+    return torch.empty_like(input, dtype=dtype)
```
Comment (manid2) [inline torch_spyre/_inductor/customops.py:460]: This custom op and CPU fallback can be removed as #1691 is merged.

Comment (imaihal) [reply torch_spyre/_inductor/customops.py:460]: CPU fallback is still required for int32–fp32 conversion, as it is not supported on Spyre.
On the other hand, since fp16–fp32 conversion is supported on Spyre by your PR #1691, CPU fallback is no longer used. Whether CPU fallback is applied is determined here:
i[f DtypeOpTable.get_operator(src_dtype, dst_dtype) is None: in .
](https://github.com/torch-spyre/torch-spyre/pull/2130/changes#diff-bed211fddc35b2fdd5f5f2a5a628b57288b6203932a447839046b9e2b2682c7fR747)

Comment (manid2) [reply torch_spyre/_inductor/customops.py:460]: ok understood.

**Review by moriohara — DISMISSED** [review]
lgtm. Confirmed this PR with int64 torch.Tensor.add and torch.mul from Q2 target models.

**Comment by moriohara** [comment]
@ani300 can you review?

File: torch_spyre/_inductor/customops.py (hunk 1 above)
Comment (ani300) [reply torch_spyre/_inductor/customops.py:460]: will #2134 fix this for real?

File: torch_spyre/_inductor/lowering.py [hunk 2]
```diff
@@ -844,3 +849,86 @@ def fill_padding(count, offset):
...
+    Args:
+        fn: The lowering function to call
+        *args: Arguments to pass to fn
+        convert_output: If True, convert output back to int64.
+                       Set to False for operations like div that should return float.
+    """
+    if not any(x.get_dtype() == torch.int64 for x in args):
+        return fn(*args)
+
+    args = [to_dtype(x, torch.float32) for x in args]
```
Comment (ani300) [inline torch_spyre/_inductor/lowering.py:888]: I'd have expected us to not do a pure conversion from int64 to fp32 but instead do some bit shuffling to ensure we are using the mantissa bits properly, something I'm missing here? Or is that what the pytorch conversion code does already?

**Review by ani300 — COMMENTED** [review]
I have 2 questions, if addressed, lgtm

File: torch_spyre/_inductor/customops.py (hunk 1 above)
Comment (moriohara) [reply torch_spyre/_inductor/customops.py:460]: https://github.com/torch-spyre/torch-spyre/issues/2156 (conversion between int32 and fp32) is supposed to remove CPU fallback

File: torch_spyre/_inductor/lowering.py (hunk 2 above)
Comment (moriohara) [reply torch_spyre/_inductor/lowering.py:888]: Here, assumed use cases don't need the full 64-bit precision, as investigated at https://github.com/torch-spyre/torch-spyre/issues/1692. The physical address arithmetic from tensor indexes DOES require full int32, for which another issue will be raised (cc @vswagath1989)

**Review by ani300 — APPROVED** [review]

**Review by avery-blanchard — APPROVED** [review]
